packages/aws-cdk/lib/api/logs/register-cloudwatch-logs.ts

```diff
@@ -0,0 +1,90 @@
+// with a specific name if one is not provided
+const RESOURCE_TYPES_WITH_IMPLICIT_LOGS = ['AWS::Lambda::Function'];
+
+export async function registerCloudWatchLogGroups(
+  sdkProvider: SdkProvider,
+  stackArtifact: cxapi.CloudFormationStackArtifact,
+): Promise<{ env: cxapi.Environment, sdk: ISDK, logGroupNames: string[] }> {
+
+
+  const resolvedEnv = await sdkProvider.resolveEnvironment(stackArtifact.environment);
+  const sdk = await sdkProvider.forEnvironment(resolvedEnv, Mode.ForReading);
+
+  const listStackResources = new LazyListStackResources(sdk, stackArtifact.stackName);
+  const evaluateCfnTemplate = new EvaluateCloudFormationTemplate({
+    stackArtifact,
+    parameters: {},
+    account: resolvedEnv.account,
+    region: resolvedEnv.region,
+    partition: (await sdk.currentAccount()).partition,
+    urlSuffix: sdk.getEndpointSuffix,
+    listStackResources,
+  });
+
+  const resources = await listStackResources.listStackResources();
+
+  const logGroupLogicalIds = new Set<string>();
+  // map of logicalId to AWS service name
+  // e.g. mylambdaFunction: lambda
+  const logicalIdsOfImplicitLogServices = new Map<string, string>();
+  //
+  // do a first pass at identifying all log groups
+  for (const resource of resources) {
+    if (resource.ResourceType === 'AWS::Logs::LogGroup') {
+      logGroupLogicalIds.add(resource.LogicalResourceId);
+      // track all the resources that will create a log group if one
+      // is not provided
+    } else if (RESOURCE_TYPES_WITH_IMPLICIT_LOGS.includes(resource.ResourceType)) {
+      const pathPart = resource.ResourceType.split('::')[1].toLowerCase();
+      logicalIdsOfImplicitLogServices.set(resource.LogicalResourceId, pathPart);
+    }
+  }
+
+  // For each log group in the template make:
+  // 1. make sure it is not associated with an excluded type
+  // 2. see if it is associated with a resource type that implicitely
+  //   creates a log group so we know not to add the implicit log group later
+  for (const id of logGroupLogicalIds) {
+    const resourcesReferencingLogGroup = evaluateCfnTemplate.findReferencesTo(id);
+    for (const r of resourcesReferencingLogGroup) {
+      if (IGNORE_LOGS_RESOURCE_TYPES.includes(r.Type)) {
+        logGroupLogicalIds.delete(id);
+      } else if (RESOURCE_TYPES_WITH_IMPLICIT_LOGS.includes(r.Type)) {
+        logicalIdsOfImplicitLogServices.delete(r.LogicalId);
+      }
+    }
+  }
+
+  const logGroupNames: string[] = [];
+  for (const id of logGroupLogicalIds) {
+    const groupName = await evaluateCfnTemplate.findPhysicalNameFor(id);
+    if (groupName) {
+      logGroupNames.push(groupName);
+    }
+  }
+
+  // some resources can be created with a custom log group (handled above).
+  // if a custom log group is not created, then the service will create one with a
+  // specific name i.e. '/aws/codebuild/project-name'
+  for (const [logicalId, serviceName] of logicalIdsOfImplicitLogServices) {
```

Comment:
We could keep track what the property that points to the physical name of each of the resources that creates implicit log groups is. For example, for `AWS::Lambda::Function`, that's `FunctionName`.

If we did that, we could use the `establishPhysicalName()` helper that I mentioned above here too.

Reply:
I updated the `RESOURCE_TYPES_WITH_IMPLICIT_LOGS` to be a map and put it there.